src/cargo/core/resolver/mod.rs

```diff
@@ -147,7 +147,10 @@ pub fn resolve(
 
     let mut cksums = HashMap::new();
     for (summary, _) in cx.activations.values() {
-        let cksum = summary.checksum().map(|s| s.to_string());
+        let cksum = summary
+            .checksum()
+            .map(|s| s.to_string())
+            .filter(|s| !s.is_empty());
```

Comment:
Not really. The problem arises from the fact that the git dependency has no checksum, so when you generate a lockfile after adding a git dependency to a project, the relevant `[[package]]` section does not contain a checksum field. When updating the dependencies, `cargo local-registry` removes the source replacement configuration before running `cargo::ops::resolve_ws`, which verifies the checksums. The way round this would be to do similar filtering in `Resolve::merge_from` instead.

For example, after https://github.com/rust-lang/cargo/blob/b690ab48fc079bf068fcf0352096ddfe5821269f/src/cargo/core/resolver/resolve.rs#L158 I added:

```rust
let mine = mine.as_ref().filter(|s| !s.is_empty());
let cksum = cksum.as_ref().filter(|s| !s.is_empty());
```

and this does appear to prevent the error from occurring. 

re lockfile stability, my main expectation with the lockfile as a user is that enabling source replacement and running a build shouldn't cause the lockfile to change. The proposed alternative change does obviously cause the lockfile to change. My usual test for whether I have set up my local registry correctly is to verify `cargo build --frozen` works and immediately after syncing the local registry, that will not work.

As far as I'm aware, only `cargo local-registry` can generate empty checksums (as git dependencies, path dependencies and vendored dependencies all generate no checksum whatsoever) so there shouldn't be another case to break compatibility with.

Reply:
I am now convinced. Thank you for explaining it again and again!

> As far as I'm aware, only `cargo local-registry` can generate empty checksums (as git dependencies, path dependencies and vendored dependencies all generate no checksum whatsoever) so there shouldn't be another case to break compatibility with.

That's also what I observed. I am going to do an FCP and see if we can reach a consensus from the team.